libs/framework/src/framework.c

```diff
@@ -190,10 +191,44 @@ celix_status_t fw_refreshHelper_stop(struct fw_refreshHelper * refreshHelper);
 struct fw_bundleListener {
 	bundle_pt bundle;
 	bundle_listener_pt listener;
+
+    celix_thread_mutex_t useMutex; //protects useCount
+    celix_thread_cond_t useCond;
+    size_t useCount;
 };
 
 typedef struct fw_bundleListener * fw_bundle_listener_pt;
 
+static inline void fw_bundleListener_destroy(fw_bundle_listener_pt listener, bool wait) {
+    celixThreadMutex_lock(&listener->useMutex);
+    while (wait && listener->useCount != 0) {
```

Comment:
It's probably better to move the wait check outside of the mutexes, unless you specifically want to synchronize on the useMutex in all cases.

Reply:
The wait is on the useCount protectedby mutex.So the lock is needed.

Reply:
Not quite. The following has exactly the same functional sementics (other than the mutex lock):
```
static inline void fw_bundleListener_destroy(fw_bundle_listener_pt listener, bool wait) {
    if(wait) {
        celixThreadMutex_lock(&listener->useMutex);
        while (listener->useCount != 0) {
            celixThreadCondition_wait(&listener->useCond, &listener->useMutex);
        }
        celixThreadMutex_unlock(&listener->useMutex);
    }

    //destroy
    celixThreadMutex_destroy(&listener->useMutex);
    celixThreadCondition_destroy(&listener->useCond);
    free(listener);
}
```

Reply:
ah yes, i was missing the point. Will update this